### PR TITLE
Added geofence to rules

### PR DESCRIPTION
**Pull request by ShuckleHQ**
## Description
Allows rules to further sub-partition filter interpretations with geofence references.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (would cause existing functionality to change)

## Motivation and Context
 Let's say you
have a "candy" filter that selects high-value candy. You want this
to notify to different channels for different parts of your city.

Attaching geofences to rules makes this easy. It also adds a number of
other options for handling multiple configurations.

## How Has This Been Tested?
Tested locally with monsters, eggs and raids

## Wiki Update

- [X] This change requires an update to the Wiki.
- [ ] This change does not require an update to the Wiki.

## Checklist
- [X] This change follows the code style of this project.
- [ ] This change only changes what is necessary to the feature.

### CODE REVIEW

File: PokeAlarm/Manager.py [hunk 1]
```diff
@@ -639,9 +641,11 @@ def process_stop(self, stop):
         rules = self.__stop_rules
         if len(rules) == 0:  # If no rules, default to all
             rules = {"default": Rule(
-                self.__stop_filters.keys(), self.__alarms.keys())}
+                self.__stop_filters.keys(), self.__alarms.keys(), None)}
 
         for r_name, rule in rules.iteritems():  # For all rules
+            if not self.check_rule_geofences(rule, stop):
+                continue
```
Comment (kvangent) [inline PokeAlarm/Manager.py]: No warning that an event was dropped?

File: PokeAlarm/Manager.py [hunk 2]
```diff
@@ -911,14 +922,7 @@ def _trigger_raid(self, raid, alarms):
...
 
-    # Check to see if a notification is within the given range
-    def check_geofences(self, f, e):
-        """ Returns true if the event passes the filter's geofences. """
-        if self.geofences is None or f.geofences is None:  # No geofences set
-            return True
-        targets = f.geofences
-        if len(targets) == 1 and "all" in targets:
-            targets = self.geofences.iterkeys()
+    def __check_fence(self, targets, e):
```
[the remark this answers is not in the document]
Comment (krosenvold) [reply PokeAlarm/Manager.py]: Didnt know that :)

**Comment by krosenvold** [comment]
Thanks for a strightforward answer. I'd just like to point out I already run managers, I scan three different cities and have a manager for each. Conceptually I find this very clean. It's when I need to subdivide the largest city into geographical regions for distinct notifications it all breaks down. So I'll end up calling all my geofences "standardfence" and create a truckload of managers if I want to achieve this goal. So I end up creating a ton of crap in my configuration, all of which creates a LOT more confusion.

I have done the updates you commented

**Comment by ShuckleHQ** [comment]
Thanks for a strightforward answer. I'd just like to point out I already run managers, I scan three different cities and have a manager for each. Conceptually I find this very clean. It's when I need to subdivide the largest city into geographical regions for distinct notifications it all breaks down. So I'll end up calling all my geofences "standardfence" and create a truckload of managers if I want to achieve this goal. So I end up creating a ton of crap in my configuration, all of which creates a LOT more confusion.

I have done the updates you commented

**Comment by mkruse02** [comment]
Love this PR!  Is it just the three redundant functions at the end holding it up?  If it's been abandoned, I could give it a go.

File: PokeAlarm/Manager.py [hunk 3]
```diff
@@ -930,6 +934,28 @@ def check_geofences(self, f, e):
                 return True
             else:  # e not in gf
                 log.debug("%s not in %s.", e.name, name)
+        return False
+
+    # Check to see if a notification is within the given range
+    def check_rule_geofences(self, rule, e):
```
[the remark this answers is not in the document]
Comment (mkruse02) [reply PokeAlarm/Manager.py]: if geofence_names were renamed to geofences or vise versa, these 3 functions could be one.

**Comment by ShuckleHQ** [comment]
I did not really abandon it, but @kvangent indicated limited enthusiasm for the kind of freedom it introduces. I hope I have made my case that this freedom is actually a very real user need that is not covered. I have so many other pokemon related things I'm doing so until I get a clearer signal its not at the top of my list. Feel free to fix the issues (you can create a pull against this pull...) or just send me the changes, I will update it :)

**Comment by mkruse02** [comment]
The use case is for ppl who have large servers with a channel for each geofence and a large number of geofences.  The current setup means that you would have to have at least 1 filter per fence/channel if not more, this could amount to 100s of filters that are essentially the same except for a different geofence.  Your PR makes it so you would only need 1 filter (or a couple) and then a rule for each fence which makes things much much much simpler, especially in terms of expanding and changing.  Just making a case for this PR :)  I'll see if I can condense the functions at the end real quick in a few hours.

**Comment by billyjbryant** [comment]
@mkruse02 the problem here becomes, once you have Geofences @ the Rules section so that multiple geofences use the same filter, you then run into the issue of people needing `custom_dts` filters for each rule.  Once you start opening the gates everyone will have a use case for why `x` should be available at the rule level.  I am not doubting the legitimate use case for `geofences` at the rule level, just playing devil's advocate of why its not the best idea.

**Comment by kvangent** [comment]
Additionally, there are some performance implications from putting too much strain into a single manager. We already have the `"geofence": [ "all" ]` shortcut and the default section. You can pretty easily specify a large number of geofences on a per manager basis, and if you are at a point where you are using hundreds of filters you might also be at the point where you are hitting one of the bounds of a single manager. One of my concerns it that if we do something like this change, it's going to lead to a bunch of people hitting limits.

Additionally, there is a lot of confusion surrounding geofences and how they work in comparison to filters (we still constantly get people using the filename instead of the geofence name), so I hesitate to add in another vector for that confusion. 

On the other hand, I certainly can emphasize with the problem. You aren't the first person (and probably wont be the last) to suggest an improvement to reduce some of the repetition between common settings in filters. I think this is the real root of the problem really - it would be nice to reduce that repetition somehow, some way to share common settings between filters. I think this feature _kind of_ touches on that, but agree with the points Billy mentioned about it being a slippery slope. If we do accept this and then later on find a solution that's more universal, it's just gonna piss everyone that was using it when I take this out and they have to update their configs. 

TL;DR: No, I haven't rejected the PR outright (If I did I would have closed it). The team and I will discuss more during a less busy week. Short term, I'm leaving it open for discussion and feedback while the team and I discuss it more  (in a week or so when we aren't prepping for a patch)

**Comment by mkruse02** [comment]
@billyjbryant The argument that I'd make as to why geofences belong at the rule level and custom_dts does not belong is that filters, alarms, and geofences all have their own dedicated files, the rules file dictates how those files interact with each other.  To me it makes more sense to have them all at the rule level than it does to leave one out.  Not trying to be argumentative for the sake of it, just trying to offer another perspective to the devil's advocate :).

@kvangent The "all" short cut is a nice one for sure, but it doesn't solve the issue of having a different channel for each geofence.  For instance, currently you could have a set up like this

"filter_1": {"min_iv": 95, "geofences": ["fence_1"]},
"filter_2": {"min_iv": 95, "geofences": ["fence_2"]}
"filter_3": {"min_iv": 95, "geofences": ["fence_3"]}

"rule_1": {"filters": "filter_1", "alarms": "alarm_1"}
"rule_2": {"filters": "filter_2", "alarms": "alarm_2"}
"rule_3": {"filters": "filter_3", "alarms": "alarm_3"}

But with this PR it can be condested:


"filter_1": {"min_iv": 95}

"rule_1": {"filters": "filter_1", "alarms": "alarm_1", "geofences": ["fence_1"]}
"rule_2": {"filters": "filter_1", "alarms": "alarm_2", "geofences": ["fence_2"]}
"rule_3": {"filters": "filter_1", "alarms": "alarm_3", "geofences": ["fence_3"]}

As for confusion and limits, I can't speak much to that except that..... And I also understand the issue that a better solution may come along and then everyone would have to change and that would suck.....  But just thought I'd add my .02 to the conversation again :)

@ShuckleHQ I PR'd your PR :)

**Comment by kvangent** [comment]
@mkruse02 The second configuration you posted could also be accomplished currently with 3 different managers, each with an alarm to a different channel.

**Comment by mkruse02** [comment]
@kvangent True, true, but once you get dozens of gfs, then you have dozens of managers and the config file gets really messy...... it's much cleaner to have dozens of rules.... but as you said before this could run into limits..... Overall, I still find the PR very useful, but I am understanding the hesitation. :)

**Comment by bbdoc** [comment]
While I agree lots can be done using managers, I think one of the main issue about managers is that it makes the configuration pretty difficult to handle when you start having lots of different managers because you are handling lots of regions and want to match geofences with discord/telegram channels. Might be worth thinking about a rewrite of the way managers are configured (maybe a managers.json file, just like you have a rules.json) which would allow people to handle like 50+ managers in a readable conf.

**Comment by kvangent** [comment]
@ShuckleHQ Apologies for sitting on this for so long, I've been pretty busy the past week and as such we haven't had time to properly discuss it among the team. 

We've decided that we aren't going to support this feature. Our reasoning is more or less what we've outline above: Overall we want to restrict filtering to the Filters, and feel that the real problem is that we need a more fine grain control over sharing settings between multiple filters than Managers seem to provide. 

We have a couple ideas in mind that should address this a little bit more, however we aren't set on a specific one and have some higher priority goals in mind for the current patch (meaning it might be a while before we add them). So If you would like to leave this PR open so that people can use it until we get the better system, I'm okay with that. The one caveat is that I'll ask that you keep it properly rebased and up to date with the master branch while it's open. If it gets too out of date, I'm going to have to close it. 

@bbdoc If you are finding the config.ini difficult to deal with at large numbers of managers, it may be easier to use the command line options via a bash script. This allows for a little more flexibility in the format of your configuration than the config.ini does.

**Comment by ShuckleHQ** [comment]
@kvangent I have rebased the PR and will continue doing this at somewhat regular intervals. I would like to make one more comment though: When you initially announced rules I had to go study the source code to understand why the heck I couldn't add geofences to a rule. Like all modelling, one choses the perspective. filters regard technical pokemon attributes, there is nothing nature-given that implies geographic location needs to be part of a filter. On the contrary I'd suggest that is where the mistake was made, it should never have been there in the first place. Making it easier to duplicate *smaller* filter files a hundreds of times will not rectify this underlying problem.  But then again, perspectives can be chosen. No matter what, I hope the underlying problem will be adressed somehow, and I certainly respect your community process on this.
